### PR TITLE
[Xamarin.Android.Build.Tasks] minimal Reference Assemblies support

### PR DESCRIPTION
**Pull request by jonathanpeppers**
This is the beginning of our support for MSBuild/Roslyn "Reference
Assemblies".

Adding the following to a `csproj` file:

    <ProduceReferenceAssembly>True</ProduceReferenceAssembly>

Causes a `bin\Debug\ref\MyLibrary.dll` to exist alongside
`bin\Debug\MyLibrary.dll`.

Two item groups, `@(ReferenceCopyLocalPaths)` and `@(ReferencePath)`
will have new `%(ReferenceAssembly)` metadata:

    bin\Debug\MyLibrary.dll
        ReferenceAssembly = C:\full\path\to\bin\Debug\ref\MyLibrary.dll

If an assembly does *not* include a reference assembly, it will still
maintain `%(ReferenceAssembly)` metadata pointing to itself:

    bin\Debug\OtherLibrary.dll
        ReferenceAssembly = C:\full\path\to\bin\Debug\OtherLibrary.dll

In some cases this metadata might not be there, such as using a path
as input to `ResolveAssemblies` (`$(OutDir)$(TargetFileName)`), so our
`ResolveAssemblies` MSBuild task should produce the value if it does
not exist.

## Changes ##

The biggest change here is to rework `ResolveAssemblies` so that it
preserves item metadata. The metadata was getting lost in various
ways:

- Use of `%(ReferencePath.Identity)` instead of `@(ReferencePath)`.
- Any new `TaskItem` instances created should use the ctor taking in
  an existing `ITaskItem`. This preserves the metadata from the
  original item.

We also needed to make some changes to support `%(ReferenceAssembly)`:

- New `TaskItem` objects created by a file path should set
  `%(ReferenceAssembly)`.
- We should check if `%(ReferenceAssembly)` is missing, and set it
  where appropriate.

I looked into reworking `_GenerateJavaStubs` so it can be skipped
when a reference assembly didn't change.

I merely changed a single input: `@(_ResolvedAssemblies)` to
`@(_ResolvedAssemblies->'%(ReferenceAssembly)')`.

Unfortunately, this would break! A developer could write this
completely valid C# class:

    internal class Example : Java.Lang.Object { }

However, since we have preserved item metadata, we can now use
`%(ReferenceAssembly)` throughout the build. We can take advantage of
it in the future.

I added some documentation about what I found out with MSBuild
"Reference Assembly" support, in general. I also added a unit test
verifying which targets skip when `$(ProduceReferenceAssembly)` is set
in a library project.

I also reworked our Xamarin.Forms integration project so that it is
netstandard instead of a shared project.

This will help us test `$(ProduceReference)` assembly, and also has a
much simpler project file!

Other changes:

- I let VS Windows fixup the `Xamarin.Android-Tests.sln file`. It
  looks like there were a few mismatched guids and other things.
- We did not have XamlC enabled:
  `[assembly: XamlCompilation (XamlCompilationOptions.Compile)]`
  this will greatly improve startup times!
- We will need to call `/t:Restore` on the Xamarin.Forms netstandard
  project now.

### CODE REVIEW

File: Documentation/guides/MSBuildReferenceAssemblies.md [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+
+In more concrete terms, assume you have a Xamarin.Forms solution
+containing a Xamarin.Forms PCL project and a referencing Android App
+project.  *Currently*, whenever the PCL project is changed, the App
+project must *always* be rebuilt.  In a
+`$(ProduceReferenceAssembly)=True` order, the App project would need
+to be rebuilt *less often*.
+
+[msbuild_refassemblies]: https://github.com/dotnet/roslyn/blob/master/docs/features/refout.md
+
```
Comment (jonathanpeppers) [inline Documentation/guides/MSBuildReferenceAssemblies.md]: The above section was the relevant good stuff from here: https://github.com/xamarin/xamarin-android/wiki/Build-Performance-Ideas#embrace-reference-assemblies

**Comment by jonpryor** [comment]
Should this also set `$(ProduceReferenceAssembly)`=True in e.g. `tests/CodeGen-Binding//Xamarin.Android.LibraryProjectZip-LibBinding/Xamarin.Android.LibraryProjectZip-LibBinding.csproj`?

**Comment by jonathanpeppers** [comment]
Hmm, that project would be good, but I don’t think anything would test an incremental build on it...

I think it might actually make more sense to convert the Xamarin.Forms integration project to use NetStandard instead of a shared project. Then enable it there.

@radekdoulik would that be ok with you? Would that hurt our plots?

**Comment by jonathanpeppers** [comment]
Well looks like calling `/t:Restore` directly isn't a good idea?
```
/Library/Frameworks/Mono.framework/Versions/5.16.0/lib/mono/msbuild/15.0/bin/MSBuild.dll /binaryLogger:/Users/builder/jenkins/workspace/xamarin-android-pr-builder-release/xamarin-android//bin/BuildRelease/msbuild-20181204T120645-prepare-external.binlog /p:Configuration=Release /p:AutoProvision=True /p:AutoProvisionUsesSudo=True /p:IgnoreMaxMonoVersion=False /p:Configuration=Release /p:AutoProvision=True /p:AutoProvisionUsesSudo=True /p:IgnoreMaxMonoVersion=False /t:Restore /v:normal Xamarin.Android-Tests.sln
...
TestRunner.xUnit.csproj : error MSB4057: The target "Restore" does not exist in the project.
```
I'm not sure why, it seemed to work for me locally?

File: Makefile [hunk 2]
```diff
@@ -37,7 +37,7 @@ all::
 	$(call MSBUILD_BINLOG,all,$(_SLN_BUILD)) $(MSBUILD_FLAGS) $(SOLUTION)
 
 all-tests::
-	MSBUILD="$(MSBUILD)" $(call MSBUILD_BINLOG,all-tests,tools/scripts/xabuild) $(MSBUILD_FLAGS) Xamarin.Android-Tests.sln
+	MSBUILD="$(MSBUILD)" $(call MSBUILD_BINLOG,all-tests,tools/scripts/xabuild) $(MSBUILD_FLAGS) Xamarin.Android-Tests.sln /restore
```
Comment (jonpryor) [inline Makefile]: I feel like this doesn't belong here, *especially* in a possible future world order (current world order?) of using `@(PackageReference)` and .NET Standard 2 (https://github.com/xamarin/java.interop/pull/393).  It feels like this should be part of `make prepare`:

```diff
diff --git a/Makefile b/Makefile
index 0d9b6646..d04a2a51 100644
--- a/Makefile
+++ b/Makefile
@@ -102,6 +102,8 @@ prepare-external:
        git submodule update --init --recursive
        nuget restore $(SOLUTION)
        nuget restore Xamarin.Android-Tests.sln
+       $(MSBUILD) /restore $(SOLUTION)
+       $(MSBUILD) /restore Xamarin.Android-Tests.sln
        (cd external/xamarin-android-tools && make prepare CONFIGURATION=$(CONFIGURATION))
        (cd $(call GetPath,JavaInterop) && make prepare CONFIGURATION=$(CONFIGURATION) JI_MAX_JDK=8)
        (cd $(call GetPath,JavaInterop) && make bin/Build$(CONFIGURATION)/JdkInfo.props CONFIGURATION=$(CONFIGURATION) JI_MAX_JDK=8)
```

File: build-tools/scripts/PrepareWindows.targets [hunk 3]
```diff
@@ -18,6 +18,7 @@
     <MSBuild Projects="$(MSBuildThisFileDirectory)..\download-bundle\download-bundle.csproj" />
     <Exec Command="$(_NuGet) restore Xamarin.Android.sln" WorkingDirectory="$(_TopDir)" />
     <Exec Command="$(_NuGet) restore Xamarin.Android-Tests.sln" WorkingDirectory="$(_TopDir)" />
+    <MSBuild Projects="Xamarin.Android-Tests.sln" Targets="Restore" />
```
Comment (jonpryor) [inline build-tools/scripts/PrepareWindows.targets]: *This* is in keeping with what I suggested [earlier](https://github.com/xamarin/xamarin-android/pull/2474/files#r239052052) (great minds think alike), though I suspect that you *also* want to call the `Restore` target on `Xamarin.Android.sln` as well.

File: build-tools/timing/timing.targets [hunk 4]
```diff
@@ -92,7 +92,7 @@
         Condition=" '%(XACaptureBuildTimingProject.Restore)' != '' "
         Command="$(_TopDir).nuget\NuGet.exe restore &quot;%(XACaptureBuildTimingProject.Restore)&quot;"
     />
-    <Exec Command="$(_XABuild) /fl /flp:LogFile=$(_OutputFile).log &quot;/logger:$(_TimingLogger);ID=$(_ID);Description=$(_Description);Commit=$(XAVersionHash);OutputPath=$(_OutputFile).xml&quot; /p:Configuration=$(Configuration) /p:AndroidSupportedAbis=x86 %(XACaptureBuildTimingProject.Identity) /t:$(_Target)" />
+    <Exec Command="$(_XABuild) /fl /flp:LogFile=$(_OutputFile).log &quot;/logger:$(_TimingLogger);ID=$(_ID);Description=$(_Description);Commit=$(XAVersionHash);OutputPath=$(_OutputFile).xml&quot; /p:Configuration=$(Configuration) /p:AndroidSupportedAbis=x86 %(XACaptureBuildTimingProject.Identity) /t:$(_Target) /restore" />
```
Comment (jonpryor) [inline build-tools/timing/timing.targets]: We need `/restore` *here* as well?

I believe what we *really* want is that *everywhere* we use `nuget restore` we should *also* call `msbuild /restore`, which would suggest:

```diff
diff --git a/build-tools/timing/timing.targets b/build-tools/timing/timing.targets
index e35c9c02..d06f26cc 100644
--- a/build-tools/timing/timing.targets
+++ b/build-tools/timing/timing.targets
@@ -92,6 +92,11 @@
         Condition=" '%(XACaptureBuildTimingProject.Restore)' != '' "
         Command="$(_TopDir).nuget\NuGet.exe restore &quot;%(XACaptureBuildTimingProject.Restore)&quot;"
     />
+    <MSBuild
+        Condition=" '%(XACaptureBuildTimingProject.Restore)' != '' "
+        Projects="%(XACaptureBuildTimingProject.Restore)"
+        Targets="Restore"
+    />
```

**Comment by jonathanpeppers** [comment]
Windows is green:
```
assets/subfolder/asset2.txt should be in the apk.
```
This is fixed in #2480

File: src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets [hunk 5]
```diff
@@ -2293,7 +2293,7 @@ because xbuild doesn't support framework reference assemblies.
 
 <Target Name="_GenerateJavaStubs"
   DependsOnTargets="_SetLatestTargetFrameworkVersion;_PrepareAssemblies;$(_AfterPrepareAssemblies)"
-  Inputs="$(MSBuildAllProjects);@(_ResolvedAssemblies);$(_AndroidManifestAbs);$(_AndroidBuildPropertiesCache);@(_AndroidResourceDest)"
+  Inputs="$(MSBuildAllProjects);@(_ResolvedAssemblies->'%(ReferenceAssembly)');$(_AndroidManifestAbs);$(_AndroidBuildPropertiesCache);@(_AndroidResourceDest)"
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets]: This optimization cannot be performed.  It's entirely valid for an assembly to declare non-`public` types which require Java Callable Wrappers:

```csharp
internal class Example : Java.Lang.Object {
}
```

`_GenerateJavaStubs` needs help, but in a different way.

**Comment by jonpryor** [comment]
Suggested Squash-and-merge commit message:

[Xamarin.Android.Build.Tasks] Start Reference Assemblies support (#2474)

```
This is the beginning of our support for MSBuild/Roslyn
"Reference Assemblies".

Adding the following to a `.csproj` file:

	<ProduceReferenceAssembly>True</ProduceReferenceAssembly>

Causes a `bin\Debug\ref\MyLibrary.dll` to be built alongside
`bin\Debug\MyLibrary.dll`.  The new `bin\Debug\ref\MyLibrary.dll`
file is *only* changed whenever *`public` and `protected`* API in
`MyLibrary.dll` is changed.

MSBuild creates two item groups, `@(ReferenceCopyLocalPaths)` and
`@(ReferencePath)`, which have `%(ReferenceAssembly)` metadata:

	bin\Debug\MyLibrary.dll
	    ReferenceAssembly = C:\full\path\to\bin\Debug\ref\MyLibrary.dll

If an assembly does *not* include a reference assembly, it will still
maintain `%(ReferenceAssembly)` metadata pointing to itself:

	bin\Debug\OtherLibrary.dll
	    ReferenceAssembly = C:\full\path\to\bin\Debug\OtherLibrary.dll

In some cases this metadata might not be there, such as using a path
as input to `<ResolveAssemblies/>` (via `$(OutDir)$(TargetFileName)`),
so our `<ResolveAssemblies/>` task will produce the value if it does
not exist.

~~ Changes ~~

The biggest change here is to rework `<ResolveAssemblies/>` so that
it preserves item metadata.  The metadata was getting lost in
various ways:

  - Use of `%(ReferencePath.Identity)` instead of `@(ReferencePath)`.
  - Any new `TaskItem` instances created should use the ctor taking
    in an existing `ITaskItem`.  This preserves the metadata from the
    original item.

We also needed to make some changes to support `%(ReferenceAssembly)`:

  - New `TaskItem` objects created by a file path should set
    `%(ReferenceAssembly)`.
  - We should check if `%(ReferenceAssembly)` is missing, and set it
    where appropriate.

I added some documentation about what I found out with MSBuild
"Reference Assembly" support, in general.  I also added a unit test
verifying which targets skip when `$(ProduceReferenceAssembly)` is
set in a library project.

To help test Reference Assembly support,
`tests/Xamarin.Forms-Performance-Integration` has been updated so a
.NET Standard 2 project is used instead of a Shared Project.

This will help us test `$(ProduceReferenceAssembly)`, and also has a
much simpler project file!

Other changes:

  - I let VS Windows fixup the `Xamarin.Android-Tests.sln file`.
    It looks like there were a few mismatched GUIDs and other things.
  - We did not have `<XamlC/>` enabled:
    `[assembly: XamlCompilation (XamlCompilationOptions.Compile)]`
    should greatly improve startup times!
```

File: src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets (hunk 5 above)
Comment (jonathanpeppers) [reply src/Xamarin.Android.Build.Tasks/Xamarin.Android.Common.targets]: Removed the changes to `_GenerateJavaStubs`...

There may yet be a target that can take advantage of `$(ProduceReferenceAssembly)` and be skipped. I will keep an eye out for one.

**Comment by MagicAndre1981** [comment]
@jonathanpeppers 

in which version is this included? In VS2017 15.9, I always get the message that the ref dll is modified even if I only change one value in XAML file.

> "not up to date as bin\Debug\ref\FooApp.dll is modified after output file"

**Comment by dellis1972** [comment]
@MagicAndre1981 which version of Xamarin.Forms are you using? I seem to remember the code behind classes of the Xaml used to be public, so changing Xaml would effect the public api. This would then cause the reference assembly to be changed. 

I *think* they might have changed that in later versions... so just make sure you are on the latest stable Xamarin.Forms and hopefully it will work.

**Comment by jonathanpeppers** [comment]
You need VS 2019 16.0 for XAML files to work properly in reference assemblies. I reported a bug for Roslyn around EmbeddedResource, they fixed and shipped in 16.0 GA.

https://github.com/dotnet/roslyn/issues/31197

**Comment by MagicAndre1981** [comment]
@jonathanpeppers thanks for the information. This explains it.

**Comment by MagicAndre1981** [comment]
ok, works in 16.3 Preview1

> Reference assembly bin\Debug\ref\Foo.App.dll" already has latest information. Leaving it untouched.

@jonathanpeppers 

please update your blogpost "Optimize your Xamarin.Android Builds" and add this information.

**Comment by jonathanpeppers** [comment]
That blog post is assuming you are using 16.2 stable. Did that not work for you?

**Comment by MagicAndre1981** [comment]
@jonathanpeppers this was not clear. I hate VS2019 because of the broken UI and so I'll stay at VS2017 and will likely skip 2019 at all. I only have a VM with Preview channel to test if UI designers repaired the broken UI mess.

**Comment by jonathanpeppers** [comment]
I added a sentence to the intro paragraph, thanks: https://devblogs.microsoft.com/xamarin/optimize-xamarin-android-builds/

**Pull request merged**